### PR TITLE
feat: add unit tests for openmldb-batch scala utils

### PR DESCRIPTION
**Pull request by zekai427**
* **Please check if the PR fulfills these requirements**

- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

try to add union test for CaseUtil

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestCaseUtil.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      StructField("FloatType", FloatType),
+      StructField("DoubleType", DoubleType),
+      StructField("DateType", DateType),
+      StructField("Timestamp", TimestampType),
+      StructField("StringType", StringType)
+    ))
+
+    val result = getYamlSchemaString(schema1)
+
+    val expectOutput = "\"BooleanType:bool\",\"ShortType:int16\",\"IntegerType:int32\",\"LongType:int64\",\"FloatType:float\"," +
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestCaseUtil.scala]: Consider using scala string("""""") to replace the escaped characters("\")

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestCaseUtil.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      StructField("FloatType", FloatType),
+      StructField("DoubleType", DoubleType),
+      StructField("DateType", DateType),
+      StructField("Timestamp", TimestampType),
+      StructField("StringType", StringType)
+    ))
+
+    val result = getYamlSchemaString(schema1)
+
+    val expectOutput = "\"BooleanType:bool\",\"ShortType:int16\",\"IntegerType:int32\"," +
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestCaseUtil.scala]: 考虑用scala triple quota(""")来避免在case中转义，这样复杂的case不太好写。

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestGraphvizUtil.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      (0, Timestamp.valueOf("1899-04-01 0:0:0"), 1.0),
+      (0, Timestamp.valueOf("1900-01-01 0:0:0"), 1.0),
+      (0, Timestamp.valueOf("1969-01-01 0:0:0"), 1.0),
+      (0, Timestamp.valueOf("2000-08-01 0:0:0"), 1.0),
+      (0, Timestamp.valueOf("2019-09-11 0:0:0"), 1.0)
+    )
+
+    val table = Session.createDataFrame(data.map(Row.fromTuple(_)).asJava, schema)
+
+    val path = getClass.getResource("").getPath + "/TestGraphvizUtil/drawPhysicalPlan.png"
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestGraphvizUtil.scala]: 这里getResource("")没有传资源可能有问题，考虑创建一个temp file来存临时文件。

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestGraphvizUtil.scala [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    val engine = new SqlEngine(sql, HybridseUtil.getDatabase("spark_db", Map("t" -> table)), engineOptions)
+    val root = engine.getPlan
+    drawPhysicalPlan(root,path)
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestGraphvizUtil.scala]: 逗号后面加空格，不然style可能不过。

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestSparkRowUtil.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.sql.{Date, Timestamp}
+
+class TestSparkRowUtil extends FunSuite{
+  test("Test createOrderKeyExtractor") {
+    val shortRow = Row.apply(1.toShort)
+    assert(createOrderKeyExtractor(0,ShortType,false).apply(shortRow) == 1L)
+
+    val intRow = Row.apply(1)
+    assert(createOrderKeyExtractor(0,IntegerType,false).apply(intRow) == 1L)
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestSparkRowUtil.scala]: assert结果应该是1，不是1L。

**Review by tobegit3hub — CHANGES_REQUESTED** [review]

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestCaseUtil.scala (hunk 2 above)
Comment (zekai427) [reply java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestCaseUtil.scala]: 已解决

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestSparkUtil.scala [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  test("Test supportNativeLastJoin") {
+    assert(!supportNativeLastJoin(JoinType.kJoinTypeFull, hasOrderby = true))
+    assert(!supportNativeLastJoin(JoinType.kJoinTypeFull, hasOrderby = false))
+    assert(!supportNativeLastJoin(JoinType.kJoinTypeLast, hasOrderby = false))
+  }
+  test("Test addIndexColumn") {
+    val Session: SparkSession = getSparkSession
+    val table: DataFrame = Session.createDataFrame(data.map(Row.fromTuple(_)).asJava, schemaTest1)
+    assert(addIndexColumn(Session,table,"test","zipwithindex")
+      .select("test").distinct().count()==6)
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestSparkUtil.scala]: 这个数字6可以考虑从上面date那里获得，否则修改data数据要改多个代码。

File: java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestSparkUtil.scala [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      StructField("id", IntegerType),
+      StructField("time", TimestampType)
+    ))
+    val schemaTest4: StructType = StructType(Seq(
+      StructField("id", IntegerType,nullable = false),
+      StructField("time", TimestampType)
+    ))
+    assert(!checkSchemaIgnoreNullable(schemaTest1,schemaTest2))
+    assert(checkSchemaIgnoreNullable(schemaTest1,schemaTest3))
+    assert(schemaTest1!=schemaTest4)
```
Comment (tobegit3hub) [inline java/openmldb-batch/src/test/scala/com/_4paradigm/openmldb/batch/utils/TestSparkUtil.scala:102]: 这个不是我们的函数，不需要assert也可以。

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Review by dl239 — APPROVED** [review]

**Review by zekai427 — COMMENTED** [review]
image
Increase from 33% to 38%

**Comment by tobegit3hub** [comment]
Previously test coverage for batch engine.

openmldb_batch_ut_coverage

**Pull request merged**